sdk/communication/AzureCommunicationChat/Source/Registrar/RegistrarClient.swift

```diff
@@ -0,0 +1,263 @@
+    ) {
+        credential.token { accessToken, error in
+            do {
+                //Get token from CommunicationTokenCredential to set the authentication header
+                guard let skypeToken = accessToken?.token else {
+                    completionHandler(
+                        nil,
+                        AzureError.client("Failed to get token from CommunicationTokenCredential.", error)
+                    )
+                    return
+                }
+                
+                let httpHeaders: HTTPHeaders = [
+                    RegistrarHeader.contentType.rawValue : RegistrarMimeType.json.rawValue,
+                    RegistrarHeader.skypeTokenHeader.rawValue : skypeToken
+                ]
+                
+                let httpRequest = try HTTPRequest(method: request.httpMethod, url: request.url, headers: httpHeaders, data: request.data)
+                
+                completionHandler(httpRequest, nil)
+            } catch {
+                let azureError = AzureError.client("Failed to add headers to the Register request.", error)
+                completionHandler(nil, azureError)
+            }
+        }
+    }
+    
+
+    /// Sends an HTTP request to Registrar.
+    /// - Parameters:
+    ///   - request: The HTTP request.
+    ///   - completionHandler: Returns the URLResponse, and an error if any errors occurred.
+    private func sendHttpRequest(
+        _ request: HTTPRequest,
+        completionHandler: @escaping (Result<HTTPResponse?, AzureError>) -> Void
+    ) {
+        let dispatchQueue = DispatchQueue.main
+        
+        let context = PipelineContext.of(keyValues: [
+            ContextKey.allowedStatusCodes.rawValue: [202] as AnyObject
+        ])
+        
+        self.request(request, context: context) { result, httpResponse in
+            switch result{
+            case .success:
+                guard (httpResponse?.statusCode) != nil else {
+                    let invalidStatusCodeError = AzureError.client("Didn't find a valid status code.")
+                    dispatchQueue.async {
+                        completionHandler(.failure(invalidStatusCodeError))
+                    }
+                    return
+                }
+                dispatchQueue.async {
+                    completionHandler(.success(httpResponse))
+                }
+            case .failure:
+                dispatchQueue.async {
+                    completionHandler(.failure(AzureError.client("Registration request failed.")))
+                }
+            }
+        }
+    }
+
+    // MARK: Internal Methods
+
+    /// Sets a registration in Registrar.
+    /// - Parameters:
+    ///   - clientDescription: RegistrarClientDescription.
+    ///   - transports: RegistrarTransports, a transport contains the APNS token as the path.
+    ///   - completionHandler: Returns the response. Success indicates the registration was received.
+    internal func setRegistration(
```

Comment:
Why are some methods `internal` and others are `private`?

Reply:
(1) Two **internal** functions "setRegistration" and "deleteRegistration" need to be accessed by another source file(ChatClient.swift) from their defining module. For example, inside chatClient.startPushNotifications(), we call registrarClient.setRegistration().  (2) The other **private** functions, such as "setRequestHeader", are designed to be called in the enclosing declaration in the same file, which is RegistrarClient.swift. These functions are used only for constructing a HTTP Request.